paddle/framework/operator.h

```diff
@@ -0,0 +1,75 @@
+/* Copyright (c) 2016 PaddlePaddle Authors. All Rights Reserve.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <boost/variant.hpp>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "paddle/framework/op_desc.pb.h"
+#include "paddle/framework/scope.h"
+#include "paddle/utils/Error.h"
+
+namespace paddle {
+namespace framework {
+
+class DeviceContext {};
+class CpuContext : public DeviceContext {};
+class GpuContext : public DeviceContext {};
+
+/**
+ * OpRunContext is the only parameter of Operator's Run function.
+ * Run will get input/output variables, state such as momentum and
+ * device resource such as CUDA stream, cublas handle, etc. from
+ * OpRunContext. User should construct it before run the Operator.
+ */
+class OpRunContext {
+ public:
+  Scope* scope;
+  DeviceContext* device_context;
+};
+
+using Attribute =
```

Comment:
I can see three ways of implementing Operator attributes:

1. `std::map<string, boost:variant<...> >`
2. `std::map<string, shared_ptr<Variable> >`
3. `OperatorDesc`, given that `OperatorBase::OperatorBase` takes a `OperatorDesc` and saves it in `desc_`.

Which is the simplest one?